### PR TITLE
major refactoring of code generator package

### PR DESCRIPTION
**Pull request by jaypipes**
The patches contained in this pull request are all stand-alone (they
may be applied individually without breaking the codebase) however
they are all small steps in a larger refactoring effort to clean up and
clarify the code generating functionality in ACK.

The `pkg/model` package is now exclusively used for structs that
*describe* the API objects and resources that are inferred from the
source API model files (in `github.com/aws/aws-sdk-go/models/apis/`).

All functionality that was outputting Go code that set SDK Input shapes
or CustomResource Spec and Status fields has been moved to a new
`pkg/generate/code` package and the previous methods like
`pkg/model:CRD.GoCodeSetInput()` have been renamed and simplified
into package-level functions to make their purpose more clear. For
example, `CRD.GoCodeSetInput()` is now `pkg/generate/code.SetSDK()`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jaypipes** [comment]
@A-Hilaly @mhausenblas @vijtrip2 @kumargauravsharma @nmvk @RedbackThomson the only effective way to review this pull request is going to be to go commit-by-commit, I'm afraid. I've tried to include descriptive commit messages for each individual commit to make the reviewing easier, but I recognize there are a lot of aggregate changes introduced here and apologize in advance for the lengthy code review.

File: pkg/model/types.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		gtwp = typeParts[0] + "." + gte
+	}
+	return gte, gt, gtwp
+}
+
+// replacePkgName accepts a type string, as returned by
+// Shape.GoTypeWithPkgName() and replaces the package name of the aws-sdk-go
+// SDK API (e.g. "ecr" for the ECR API) with the string "svcsdkapi" which is
+// the only alias we always use in our templated output.
+func replacePkgName(
```
Comment (RedbackThomson) [inline pkg/model/types.go]: minor nit: I still don't necessarily understand the effect of this operation, even after reading the comment.

File: pkg/model/crd_test.go [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
 
 import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+
+	"github.com/aws/aws-controllers-k8s/pkg/model"
 )
 
 func TestReplacePkgName(t *testing.T) {
```
Comment (A-Hilaly) [inline pkg/model/crd_test.go:11]: nit: we can move this unit test to a `types_test.go` file, since `ReplacePkgName` is defined in `types.go`

**Review by A-Hilaly — APPROVED** [review]
Thank you for splitting the changes into multiple commits @jaypipes, it made it very easy to review.
/lgtm 👍

File: pkg/model/crd_test.go (hunk 2 above)
Comment (jaypipes) [reply pkg/model/crd_test.go:11]: yup. done.

File: pkg/model/types.go (hunk 1 above)
Comment (jaypipes) [reply pkg/model/types.go]: :) I will try to fix this up to be more understandable.

File: pkg/generate/code/check.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	return ""
+}
+
+// CheckRequiredFieldsMissingFromShape returns Go code that contains a
+// condition checking that the required fields in the supplied Shape have a
+// non-nil value in the corresponding CR's Spec or Status substruct.
+//
+// Sample Output:
+//
+// return r.ko.Spec.APIID == nil || r.ko.Status.RouteID != nil
```
Comment (vijtrip2) [inline pkg/generate/code/check.go:58]: correction in comment: This should be `return r.ko.Spec.APIID == nil || r.ko.Status.RouteID == nil`

**Review by vijtrip2 — APPROVED** [review]

File: pkg/generate/code/check.go (hunk 3 above)
Comment (jaypipes) [reply pkg/generate/code/check.go]: fixed.

File: pkg/generate/apigwv2_test.go [hunk 4]
```diff
@@ -184,77 +183,4 @@ func TestAPIGatewayV2_Route(t *testing.T) {
...
-		ko.Status.APIGatewayManaged = resp.ApiGatewayManaged
-	}
-	if resp.RouteId != nil {
-		ko.Status.RouteID = resp.RouteId
-	}
-`
-	assert.Equal(expCreateOutput, crd.GoCodeSetOutput(model.OpTypeCreate, "resp", "ko", 1, false))
-
-	expReadOneOutput := `
-	if resp.ApiGatewayManaged != nil {
```
Comment (muvaf) [inline pkg/generate/apigwv2_test.go:260]: I see lots of tests removed. Was that intentional? Sometimes I remove them until the code gets ready and forget to add them back, so just wanted to check :)

**Review by muvaf — APPROVED** [review]
LGTM! I loved the clarifications, thanks!

File: pkg/generate/apigwv2_test.go (hunk 4 above)
Comment (jaypipes) [reply pkg/generate/apigwv2_test.go:260]: @muvaf nope, I just moved the tests to different files. The previous `GoCodeSetInput` checks were moved to `pkg/generate/code/set_sdk_test.go`. `GoCodeSetOutput` were moved to `pkg/generate/code/set_resource_test.go`, etc... :)

**Pull request merged**